### DISCUSSION TITLE
【Request help】How do I use these module (like GlutWindow,AndroidWindow,OverlayWindow,Oxy theme etc.)?

### DISCUSSION THREAD

**Discussion by c61292558** [fltk-rs/fltk-rs#1362] [Q&A] [answered]
Hi,I am new in Rust Fltk-rs.
(Linux platform.A distribution system named Deepin20.8 based on Debian)
I am looking the docs about Fltk-rs.(https://docs.rs/fltk/latest/fltk/)
Now  I'm having some problems:

### 1.In the Module Window,I want to try these windows:(https://docs.rs/fltk/latest/fltk/window/index.html)
**So I typed the codes and then report errors:**

```
[package]
name = "fltk_test"
version = "0.1.0"
edition = "2021"

[dependencies]

fltk = "1.3.31"
fltk-theme="0.6.0"
fltk-table = "0.2.1"
```

-------------------------------------------------------------------------------------------------------
```
extern crate fltk_table;
use fltk::{prelude::*, *};
use fltk_theme::{widget_themes,WidgetScheme, WidgetTheme,SchemeType,ThemeType};
// use crate::frame::Frame;
use fltk::frame::Frame;
use fltk::enums::Color;
use fltk::image::SvgImage;
use fltk::{    app, enums,    window};
use fltk_table::{SmartTable, TableOpts};
// use fltk_sys::window::*;
// use crate::enums::FrameType::DownBox;
use fltk::enums::FrameType::*;

fn main() {
    let a = app::App::default();
    let widget_theme = WidgetTheme::new(ThemeType::Oxy);//^^^ variant or associated item not found in `ThemeType`
    widget_theme.apply();
    let mut win =  window::GlutWindow::default().with_size(500, 500);
    // win.set_color(Color::from_hex(0xf5f5dc));
    
    let mut btn = button::Button::new(1000, 200, 180, 130, "Hello");
    let mut btn2 = button::RadioButton::new(0, 0, 80, 30, "Hello2");
    let mut btn3 = button::CheckButton::new(80, 0, 80, 30, "Hello3");
    let mut btn4 = button::LightButton::new(160, 0, 80, 30, "Hello4");
    let mut btn5 = button::RepeatButton::new(240, 0, 80, 30, "Hello5");
    let mut btn6 = button::RadioLightButton::new(320, 0, 80, 30, "Hello6");
    let mut btn7 = button::RadioRoundButton::new(0, 100, 80, 30, "Hello7");
    let mut btn8 = button::ShortcutButton::new(80, 180, 180, 30, "shortcutbutton");
    
    let mut frame = Frame::new(0, 200, 80, 30, "Frame test");
    frame.set_color(Color::Blue);

    let mut choice = menu::Choice::default().with_size(80, 30).center_of_parent().with_label("Select item");
    choice.add_choice("1 Vue");
    choice.add_choice("2 React");
    choice.add_choice("3 javascript+jQuery");

    let flex = group::Flex::default().with_size(100, 100).column().center_of_parent();
    let mut input = input::IntInput::default();
    
    btn.set_down_frame(enums::FrameType::DownBox);//Why is the error reported when I use 'FrameType::DownBox' instead ?

    win.end();
    win.show();
 
    a.run().unwrap();
    println!("hello world",);
}
```

```   
error[E0433]: failed to resolve: could not find `GlutWindow` in `window`
  --> src/main.rs:18:28
      |
18 |     let mut win =  window::GlutWindow::default().with_size(500, 500);
     |                          ^^^^^^^^^^
    |                           |
    |                            could not find `GlutWindow` in `window`
    |                            help: a struct with a similar name exists: `Fl_Window` 
```
### 2. I can see the theme "Oxy" in this docs url(https://docs.rs/fltk/1.3.31/fltk/app/enum.Scheme.html).
**But why do it reporte error?**
```
error[E0599]: no variant or associated item named `Oxy` found for enum `ThemeType` in the current scope
  --> src/main.rs:16:52
   |
16 |     let widget_theme = WidgetTheme::new(ThemeType::Oxy);//^^^ variant or associated item not found in `ThemeType`
   |                                                    ^^^ variant or associated item not found in `ThemeType`

```
Docs shows:
```
#[non_exhaustive]
pub enum Scheme {
    Base,
    Plastic,
    Gtk,
    Gleam,
    Oxy,
}
```

### 3.`button.set_down_frame(enums::FrameType::DownBox);//Why is the error reported when I use 'FrameType::DownBox' instead ?`
I want to know the `set_down_frame()` function and wonder the effect but I seems nothing happened.

**Thanks for help.**

**Comment by MoAlyousef** [comment, marked as answer]
1- GlutWindow and GlWindow require the `enable-glwindow` flag:
```toml
[dependencies]
fltk = { version = "1.3.31", features=["enable-glwindow"] }
```

2- Oxy is part of the fltk-rs schemes, and can be used like:
```rust
let a = app::App::default().with_scheme(app::Scheme::Oxy);
```
The themes that are part of fltk-theme can be found here:
https://docs.rs/fltk-theme/latest/fltk_theme/enum.ThemeType.html

3- The down_frame is what you see when you press the button.
If you want to use `enums::FrameType::DownBox` you need to import `fltk::enums` or just `fltk::*`.
if you want to use `FrameType::DownBox` you need to import `fltk::enums::FrameType`. This is a question of how to do basic imports in Rust, and is not really related to fltk-rs.

**Reply by c61292558** [reply]
Thanks for your help.
1-But when to use this syntax `features=["enable-glwindow"]` next time or the other situation？
(By the way ,why my window don't have **'maximize'** button in the upper right corner？)
3-Excuse me,can you give some example code about the `set_down_frame` so that can show the effect ? I can't see anything effect.

**Reply by MoAlyousef** [reply]
1- In your Cargo.toml. 

2- if you remove the theme code and use `button.set_down_frame(FrameType::UpBox)` you would notice that when you press the button, it would seem as though it wasn’t clicked since the default frame is also UpBox.

**Reply by c61292558** [reply]
thank you very much.
1-**In which** case the **feature** must be added in this way?
4-By the way,how can I made the window have the  'maximize' button in the upper right corner？

**Reply by MoAlyousef** [reply]
1- your Cargo.toml needs to include this:
```toml
[dependencies]
fltk = { version = "1.3.31", features=["enable-glwindow"] }
```
In addition to any other dependencies you need. Please refer to this: https://doc.rust-lang.org/cargo/reference/features.html#dependency-features

2- you need to make the window resizable:
`window.make_resizable(true);`

**Reply by c61292558** [reply]
Thanks again.
2-I use the `window.make_resizable(true);`,now I have a new problem:all the widget were scaled.How to solve it?

**Reply by MoAlyousef** [reply]
Then instead of make_resizable, choose a single widget that you would like to be resizable along with the window `window.resizable(&some_widget);`. If you want nothing to scale, place every widget inside a Group widget and make that group widget the sole child of your window.

**Reply by c61292558** [reply]
Oh,I see! Thank you very much.
See light suddenly.

**Reply by MoAlyousef** [reply]
No problem 😉
